### PR TITLE
Publicize `ExtensionEntry`

### LINKED ISSUE #294: Trainer: publicize ExtensionEntry
Allow `manager.extend(extension_entry)` so that users can specify custom trigger etc.

### PR DESCRIPTION
**Pull request by asi1024**
Fix #294

https://github.com/pfnet/pytorch-pfn-extras/pull/304#issuecomment-911885380

### CODE REVIEW

File: pytorch_pfn_extras/training/extension.py [hunk 1]
```diff
@@ -222,3 +224,32 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
 
 def _as_extension(ext: 'ExtensionLike') -> Extension:
     return ext if isinstance(ext, Extension) else _WrappedExtension(ext)
+
+
+class ExtensionEntry:
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:229]: Please add docstring and add to Sphinx docs.

Comment (asi1024) [reply pytorch_pfn_extras/training/extension.py:229]: Added docstring. PTAL!

Comment (kmaehashi) [reply pytorch_pfn_extras/training/extension.py:229]: Could you add to Sphinx docs as well?

Comment (asi1024) [reply pytorch_pfn_extras/training/extension.py:229]: Fixed.

File: pytorch_pfn_extras/training/extension.py [hunk 2]
```diff
@@ -222,3 +224,48 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
...
+            Default is ``False``.
+
+    .. seealso::
+       :meth:`pytorch_pfn_extras.training.ExtensionsManager.extend`
+    """
+
+    def __init__(
+            self,
+            extension: Extension,
+            name: Optional[str] = None,
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:250]: Please make it keyword only arguments considering the usecase. (in most cases users want to specify triggers)

File: pytorch_pfn_extras/training/extension.py [hunk 3]
```diff
@@ -222,3 +224,48 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
...
+            priority: int = Extension.priority,
+            trigger: 'TriggerLike' = Extension.trigger,
+            call_before_training: bool = False,
+    ) -> None:
+        self.name = name or extension.name
+        self.extension = extension
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:256]: Please check that given instance is an Extension so that users do not pass callable here.

File: pytorch_pfn_extras/training/manager.py [hunk 4]
```diff
@@ -340,6 +317,11 @@ def extend(
         if self._start_extensions_called:
             raise RuntimeError(
                 'extend called after the extensions were initialized')
+
+        if isinstance(extension, extension_module.ExtensionEntry):
+            self._extensions[extension.name] = extension
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:322]: Please take care of duplicate extension name case.

File: tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        loss = F.l1_loss(y, t)
+        ppe.reporting.report({prefix + '/loss': loss})
+        return loss
+
+
+@pytest.mark.parametrize('device', ['cpu', 'cuda'])
+def test_trainer(device, path):
+    model = MyModel()
+    ppe.to(model, device)
+    model_with_loss = MyModelWithLossFn(model)
```
Comment (kmaehashi) [inline tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py:55]: Why not just transfer `model_with_loss` with `ppe.to`?

File: tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+
+    def forward(self, x, t):
+        y = self.model(x)
+        prefix = 'train' if self.training else 'val'
+        loss = F.l1_loss(y, t)
+        ppe.reporting.report({prefix + '/loss': loss})
+        return loss
```
Comment (kmaehashi) [inline tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py:48]: These model definitions can be extracted out as it's shared by many tests?

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py (hunk 6 above)
Comment (asi1024) [reply tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py:48]: Why not refactor test codes in another PR? I will open a new issue.

**Comment by asi1024** [comment]
PTAL.

File: pytorch_pfn_extras/training/extension.py [hunk 7]
```diff
@@ -222,3 +224,52 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
...
+       :meth:`pytorch_pfn_extras.training.ExtensionsManager.extend`
+    """
+
+    def __init__(
+            self,
+            extension: 'ExtensionLike',
+            *,
+            name: Optional[str] = None,
+            priority: int = Extension.priority,
+            trigger: 'TriggerLike' = Extension.trigger,
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:253]: The default value of `priority` and `trigger` needs to be provided by an extension, like `name`.

File: pytorch_pfn_extras/training/extension.py [hunk 8]
```diff
@@ -222,3 +224,52 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
...
 def _as_extension(ext: 'ExtensionLike') -> Extension:
     return ext if isinstance(ext, Extension) else _WrappedExtension(ext)
+
+
+class ExtensionEntry:
+    """Extension and options.
+
+    Args:
+        extension: An extension.
+        name: Name of extension. If omitted, inferred from the attributes of
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:234]:
```suggestion
        name: Name of extension. If omitted, copied from the attributes of
```

File: pytorch_pfn_extras/training/manager.py [hunk 9]
```diff
@@ -340,6 +317,17 @@ def extend(
...
             raise RuntimeError(
                 'extend called after the extensions were initialized')
+
+        if isinstance(extension, extension_module.ExtensionEntry):
+            if name is not None:
+                raise ValueError('"name" is duplicated')
+            if trigger is not None:
+                raise ValueError('"trigger" is duplicated')
+            if priority is not None:
+                raise ValueError('"prioirity" is duplicated')
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:327]: I think the new values should override the given ExtensionEntry attributes.
(Also take care of `call_before_training`)

File: pytorch_pfn_extras/training/manager.py [hunk 10]
```diff
@@ -349,20 +337,26 @@ def extend(
 
         if trigger is None:
             trigger = ext.trigger
-        trigger = trigger_module.get_trigger(trigger)
 
         if priority is None:
             priority = ext.priority
 
+        self._extend(
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:344]: The normalizations above (line 331 to 343) can be done in ExtensionEntry.

File: tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py (hunk 5 above)
Comment (kmaehashi) [reply tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py:55]: I think `ppe.to(model, device)` is not needed?

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: pytorch_pfn_extras/training/manager.py (hunk 10 above)
Comment (asi1024) [reply pytorch_pfn_extras/training/manager.py:344]: Normalization operation requires `self._extensions`.

**Comment by asi1024** [comment]
PTAL.

File: pytorch_pfn_extras/training/extension.py [hunk 11]
```diff
@@ -231,7 +231,7 @@ class ExtensionEntry:
 
     Args:
         extension: An extension.
-        name: Name of extension. If omitted, inferred from the attributes of
+        name: Name of extension. If omitted, copied from the attributes of
             ``extension`` argument.
         priority: Invocation priority of the extension.
             Default is ``PRIORITY_READER``.
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:237]: Please update docstrings of default values for priority, trigger.

File: pytorch_pfn_extras/training/extension.py [hunk 12]
```diff
@@ -222,3 +224,60 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
...
+       :meth:`pytorch_pfn_extras.training.ExtensionsManager.extend`
+    """
+
+    def __init__(
+            self,
+            extension: 'ExtensionLike',
+            *,
+            name: Optional[str] = None,
+            priority: Optional[int] = None,
+            trigger: Optional['TriggerLike'] = None,
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:253]: I think it's nice to use `Optional['TriggerLike']` here, so could you remove `Optional` from TriggerLike definition?
https://github.com/pfnet/pytorch-pfn-extras/blob/a8d02a665de0dfaa6f142a065d7ba6ab5c13b41d/pytorch_pfn_extras/training/_trigger_util.py

File: pytorch_pfn_extras/training/manager.py [hunk 13]
```diff
@@ -340,29 +317,27 @@ def extend(
...
-            name = ext.name or ext.default_name
-        if name == 'training':
-            raise ValueError(
-                'the name "training" is prohibited as an extension name')
 
-        if trigger is None:
-            trigger = ext.trigger
-        trigger = trigger_module.get_trigger(trigger)
+        if not isinstance(extension, extension_module.ExtensionEntry):
+            extension = extension_module.ExtensionEntry(extension)
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/manager.py:322]:
```suggestion
        if isinstance(extension, extension_module.ExtensionEntry):
            entry = extension
        else:
            entry = extension_module.ExtensionEntry(extension)
```

File: pytorch_pfn_extras/training/extension.py [hunk 14]
```diff
@@ -222,3 +224,60 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
...
+            call_before_training: bool = False,
+    ) -> None:
+        self.extension = _as_extension(extension)
+        self.priority = priority or self.extension.priority
+        self.call_before_training = call_before_training
+
+        self.set_trigger(trigger or self.extension.trigger)
+        self.set_name(name or self.extension.name or self.extension.default_name)
+
+    def set_trigger(self, trigger: 'TriggerLike') -> None:
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:263]: Please make it private as this is a public class.
```suggestion
    def _update_trigger(self, trigger: 'TriggerLike') -> None:
```

File: pytorch_pfn_extras/training/extension.py [hunk 15]
```diff
@@ -222,3 +224,60 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
...
+        self.trigger = _trigger_util.get_trigger(trigger)
+
+    def set_name(self, name: str) -> None:
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:266]:
```suggestion
    def _update_name(self, name: str) -> None:
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: pytorch_pfn_extras/training/extension.py (hunk 12 above)
Comment (asi1024) [reply pytorch_pfn_extras/training/extension.py:253]: `ExtensionEntry.__init__` interprets `trigger=None` as omitted option, whereas `get_trigger` interprets it as `never_trigger`.

Comment (kmaehashi) [reply pytorch_pfn_extras/training/extension.py:253]: Ah, I see, thanks for clarification 😄

File: pytorch_pfn_extras/training/extension.py [hunk 16]
```diff
@@ -222,3 +224,60 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
 
 def _as_extension(ext: 'ExtensionLike') -> Extension:
     return ext if isinstance(ext, Extension) else _WrappedExtension(ext)
+
+
+class ExtensionEntry:
+    """Extension and options.
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:230]:
```suggestion
    """Extension and options.
    
    When name, priority, or trigger is not specified, it is copied from the
    attributes of the given ``extension``.
```

File: pytorch_pfn_extras/training/extension.py [hunk 17]
```diff
@@ -222,3 +224,60 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
...
+    Args:
+        extension: An extension.
+        name: Name of extension. If omitted, copied from the attributes of
+            ``extension`` argument.
+        priority: Invocation priority of the extension.
+            Default is ``PRIORITY_READER``.
+        trigger: Trigger object that determines when to invoke the extension.
+            Default is ``(1, 'iteration')``.
+        call_before_training: Flag to call extension before training.
+            Default is ``False``.
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:241]:
```suggestion
        extension: An extension.
        name: Name of extension.
        priority: Invocation priority of the extension.
        trigger: Trigger object that determines when to invoke the extension.
        call_before_training: Flag to call extension before training.
```

File: tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py [hunk 18]
```diff
@@ -0,0 +1,73 @@
+import tempfile
```
Comment (kmaehashi) [inline tests/pytorch_pfn_extras_tests/training_tests/test_extension_entry.py:1]: * As a *unit* test, please use ExtensionsManager instead of Trainer.
* Please try to minimize the code, I think `MyModel` or `MyModelWithLossFn` is too much to test ExtensionEntry.
* Please make sure that extensions are correctly registered (attributes are overridden as expected) and invoked as intended.

**Review by kmaehashi — CHANGES_REQUESTED** [review]

**Review by kmaehashi — APPROVED** [review]
Thank you, LGTM!

File: pytorch_pfn_extras/training/extension.py [hunk 19]
```diff
@@ -222,3 +224,59 @@ def decorator(ext: 'ExtensionLike') -> 'ExtensionLike':
 
 def _as_extension(ext: 'ExtensionLike') -> Extension:
     return ext if isinstance(ext, Extension) else _WrappedExtension(ext)
+
+
+class ExtensionEntry:
+    """Extension and options.
+
```
Comment (kmaehashi) [inline pytorch_pfn_extras/training/extension.py:231]: flake8
```suggestion

```
